### PR TITLE
Map MODS geo extension to COCINA

### LINKED ISSUE #1124: Map MODS geo extension to COCINA
https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/mods_cocina_mappings/mods_to_cocina_geo_extension.txt

### PR DESCRIPTION
**Pull request by aaron-collier**
## Why was this change made?

Fixes #1124 

## How was this change tested?

Unit tests

### CODE REVIEW

**Comment by justinlittman** [comment]
Well, that didn't go according to plan:
```
Error: #/components/schemas/Description/properties/geographic expected array, but received ActiveSupport::HashWithIndifferentAccess: {"form"=>[{"value"=>"application/x-esri-shapefile", "type"=>"media type", "source"=>{"value"=>"IANA media type terms"}}, {"value"=>"Shapefile", "type"=>"data format"}, {"value"=>"Dataset#Point", "type"=>"type"}], "subject"=>[{"structuredValue"=>[{"value"=>"94.44342", "type"=>"west"}, {"value"=>"32.208443", "type"=>"south"}, {"value"=>"102.806404", "type"=>"east"}, {"value"=>"38.174961", "type"=>"north"}], "type"=>"bounding box coordinates", "encoding"=>{"value"=>"decimal"}, "standard"=>{"code"=>"EPSG:4326", "type"=>"coordinate reference system"}}]} (1 errors)
Examples: druid:hh789kw8606
Error: #/components/schemas/Description/properties/geographic expected array, but received ActiveSupport::HashWithIndifferentAccess: {"form"=>[{"value"=>"image/jpeg", "type"=>"media type", "source"=>{"value"=>"IANA media type terms"}}, {"value"=>"Image", "type"=>"media type", "source"=>{"value"=>"DCMI Type Vocabulary"}}], "subject"=>[{"structuredValue"=>[{"value"=>"41.888817", "type"=>"latitude"}, {"value"=>"12.464595", "type"=>"longitude"}], "type"=>"point coordinates", "encoding"=>{"value"=>"decimal"}}]} (1 errors)
Examples: druid:pk887cc4056
Error: #/components/schemas/Description/properties/geographic expected array, but received ActiveSupport::HashWithIndifferentAccess: {"form"=>[{"value"=>"application/x-esri-shapefile", "type"=>"media type", "source"=>{"value"=>"IANA media type terms"}}, {"value"=>"Shapefile", "type"=>"data format"}, {"value"=>"Dataset#Polygon", "type"=>"type"}], "subject"=>[{"structuredValue"=>[{"value"=>"3.688636", "type"=>"west"}, {"value"=>"45.231392", "type"=>"south"}, {"value"=>"4.759944", "type"=>"east"}, {"value"=>"46.275913", "type"=>"north"}], "type"=>"bounding box coordinates", "encoding"=>{"value"=>"decimal"}, "standard"=>{"code"=>"EPSG:4326", "type"=>"coordinate reference system"}}]} (1 errors)
Examples: druid:dn917yr9650
```
+ many more.

**Comment by aaron-collier** [comment]
@justinlittman not seeing the above errors any longer. My branch reported:

```
4123 of 350000 (1.178%)
Error: #/components/schemas/Standard does not define properties: type (2585 errors)
Examples: druid:zw685dw6356, druid:hv751kj0082, druid:ft545jn7796, druid:pf602td9391, druid:hv022gv3986, druid:wt776rw3893, druid:tt596nz1569, druid:mn953qz8474, druid:mj707sx9198, druid:jr037zj8458
Error: key not found: "1998-177" (545 errors)
Examples: druid:qj052dg7139, druid:bt618kd4089, druid:bs336kd7628, druid:cn644rh0549, druid:hb437sq0624, druid:rh259jf6384, druid:xv702pg0147, druid:pm039rw6886, druid:zz259fj8366, druid:wt205gq3582
Error: key not found: nil (499 errors)
Examples: druid:fj803by4391, druid:xv328kn1721, druid:zz239fk6026, druid:sh496df9903, druid:vz782dr8394, druid:pm049nq1021, druid:zc423zy5751, druid:sb459dw7958, druid:hz750nt9912, druid:qk295sk1864
Error:  isn't one of in #/components/schemas/DescriptiveBasicValue/allOf/2/properties/value (77 errors)
Examples: druid:jp270pd4425, druid:rk160wd1685, druid:tq048zj2285, druid:cg786yh6695, druid:nf221qh5784, druid:dg875gq3366, druid:xj910rk4974, druid:zx851yx8537, druid:qj331tp7537, druid:kd511sf2357
Error: key not found: "isReferencedby"
Did you mean?  "isReferencedBy" (71 errors)
Examples: druid:nf343mj2692, druid:zr238yt3712, druid:nh939rq0328, druid:yk983tq9075, druid:xm721dj4603, druid:zn537zy2368, druid:mk506vd5013, druid:nm455rd6672, druid:bq134cn1846, druid:qj046wz7225
Error: key not found: "#N/A" (64 errors)
Examples: druid:yx761xn2403, druid:py805hz3357, druid:xc343yp1189, druid:ww045zr3200, druid:wv546jr1527, druid:pm196jv8522, druid:wr549wq1857, druid:yq236gr6442, druid:pf243yr7799, druid:yd417nx0007
Error: #/components/schemas/Source/properties/code does not allow null values (56 errors)
Examples: druid:dv648jc0542, druid:bv130zg8994, druid:dz692fn7184, druid:pc866nj8061, druid:vv180sr0636, druid:sg578sb1573, druid:gd289np1088, druid:pc434cq1315, druid:ms180kr4831, druid:nw905rb5515
Error: key not found: "Corporate" (55 errors)
Examples: druid:mf809yr5431, druid:mb538bh7915, druid:zq373kq8149, druid:mb519mb0200, druid:jn504xh0935, druid:bk645jn7120, druid:jn205vq6197, druid:jn179tr9029, druid:fc016xq0501, druid:mg831sp4915
Error: name not found for 1 (53 errors)
Examples: druid:dr890hr3633, druid:dv769rz7499, druid:sh886pg5351, druid:mj684vg5303, druid:py868fw9686, druid:yf489wf7907, druid:qp044xy3548, druid:dd221kq0052, druid:jx476qg0951, druid:pb046nh9306
Error: key not found: "" (36 errors)
Examples: druid:bv906dj0910, druid:ft228gt4019, druid:fq254hh2957, druid:ff265xb5816, druid:dh545kq3183, druid:ct356kk5797, druid:zx867rz8188, druid:dx835nd1155, druid:pk902jf7671, druid:hk474mp2227
Error: #/components/schemas/Standard/properties/code does not allow null values (33 errors)
Examples: druid:zb236zf6024, druid:tx035nd3189, druid:bb982gs3824, druid:xv162nt8601, druid:xx413hy0767, druid:qy092ps2157, druid:qf043nx6278, druid:qj293wm4063, druid:nc912xp1840, druid:mj221qh8338
Error: ./mods:role/mods:roleTerm[@type="code"] is missing required authority attribute (20 errors)
Examples: druid:cr391gq7466, druid:ds919sg5985, druid:jq959xy6710, druid:nf676xt4970, druid:jn867nx7451, druid:nc018xn7579, druid:tr882hg3507, druid:tr870pd7774, druid:zc221rc0493, druid:ds846xd1044
Error: key not found: "Topic" (11 errors)
Examples: druid:xv661kx4223, druid:qr684zq1130, druid:xy743kk5793, druid:yw448qd5583, druid:ky082np9069, druid:gf424bq2099, druid:rd235fv8920, druid:fz099qm8777, druid:bq782pp3515, druid:jx066kr2212
Error: key not found: "Personal" (11 errors)
Examples: druid:rr760kw8077, druid:sd106bj5619, druid:jc292hs6731, druid:fb460yd3862, druid:jv715vs5567, druid:dr978cj5855, druid:sd744qc5142, druid:mg668kk8822, druid:zj739jx1550, druid:wd641rc1638
Error: key not found: "primary" (3 errors)
Examples: druid:rw667tg8009, druid:md049mq5771, druid:wq793rc8271
Error: key not found: "topic" (1 errors)
Examples: druid:kd333yz5798
Error: key not found: "naf" (1 errors)
Examples: druid:pw260mt1867
Error: key not found: "personal" (1 errors)
Examples: druid:sm835nh3655
Error: key not found: "lcsh" (1 errors)
Examples: druid:xw235wd2174
```

Which I don't think are related to my branch.

**Comment by justinlittman** [comment]
I would expect less. Can you rebase on master and try again?

**Review by justinlittman — APPROVED** [review]
Approve pending test.

**Comment by aaron-collier** [comment]
Yeah - a bit less when rebased:

```
3630 of 350000 (1.0371428571428571%)
Error: #/components/schemas/Standard does not define properties: type (2585 errors)
Examples: druid:zw685dw6356, druid:hv751kj0082, druid:ft545jn7796, druid:pf602td9391, druid:hv022gv3986, druid:wt776rw3893, druid:tt596nz1569, druid:mn953qz8474, druid:mj707sx9198, druid:jr037zj8458
Error: key not found: "1998-177" (545 errors)
Examples: druid:qj052dg7139, druid:bt618kd4089, druid:bs336kd7628, druid:cn644rh0549, druid:hb437sq0624, druid:rh259jf6384, druid:xv702pg0147, druid:pm039rw6886, druid:zz259fj8366, druid:wt205gq3582
Error:  isn't one of in #/components/schemas/DescriptiveBasicValue/allOf/2/properties/value (83 errors)
Examples: druid:jp270pd4425, druid:rk160wd1685, druid:tq048zj2285, druid:cg786yh6695, druid:nf221qh5784, druid:dg875gq3366, druid:fd184wn0356, druid:xj910rk4974, druid:zx851yx8537, druid:qj331tp7537
Error: key not found: "isReferencedby"
Did you mean?  "isReferencedBy" (71 errors)
Examples: druid:nf343mj2692, druid:zr238yt3712, druid:nh939rq0328, druid:yk983tq9075, druid:xm721dj4603, druid:zn537zy2368, druid:mk506vd5013, druid:nm455rd6672, druid:bq134cn1846, druid:qj046wz7225
Error: key not found: "#N/A" (64 errors)
Examples: druid:yx761xn2403, druid:py805hz3357, druid:xc343yp1189, druid:ww045zr3200, druid:wv546jr1527, druid:pm196jv8522, druid:wr549wq1857, druid:yq236gr6442, druid:pf243yr7799, druid:yd417nx0007
Error: #/components/schemas/Source/properties/code does not allow null values (56 errors)
Examples: druid:dv648jc0542, druid:bv130zg8994, druid:dz692fn7184, druid:pc866nj8061, druid:vv180sr0636, druid:sg578sb1573, druid:gd289np1088, druid:pc434cq1315, druid:ms180kr4831, druid:nw905rb5515
Error: key not found: "Corporate" (55 errors)
Examples: druid:mf809yr5431, druid:mb538bh7915, druid:zq373kq8149, druid:mb519mb0200, druid:jn504xh0935, druid:bk645jn7120, druid:jn205vq6197, druid:jn179tr9029, druid:fc016xq0501, druid:mg831sp4915
Error: name not found for 1 (53 errors)
Examples: druid:dr890hr3633, druid:dv769rz7499, druid:sh886pg5351, druid:mj684vg5303, druid:py868fw9686, druid:yf489wf7907, druid:qp044xy3548, druid:dd221kq0052, druid:jx476qg0951, druid:pb046nh9306
Error: key not found: "" (36 errors)
Examples: druid:bv906dj0910, druid:ft228gt4019, druid:fq254hh2957, druid:ff265xb5816, druid:dh545kq3183, druid:ct356kk5797, druid:zx867rz8188, druid:dx835nd1155, druid:pk902jf7671, druid:hk474mp2227
Error: #/components/schemas/Standard/properties/code does not allow null values (33 errors)
Examples: druid:zb236zf6024, druid:tx035nd3189, druid:bb982gs3824, druid:xv162nt8601, druid:xx413hy0767, druid:qy092ps2157, druid:qf043nx6278, druid:qj293wm4063, druid:nc912xp1840, druid:mj221qh8338
Error: ./mods:role/mods:roleTerm[@type="code"] is missing required authority attribute (20 errors)
Examples: druid:cr391gq7466, druid:ds919sg5985, druid:jq959xy6710, druid:nf676xt4970, druid:jn867nx7451, druid:nc018xn7579, druid:tr882hg3507, druid:tr870pd7774, druid:zc221rc0493, druid:ds846xd1044
Error: key not found: "Topic" (11 errors)
Examples: druid:xv661kx4223, druid:qr684zq1130, druid:xy743kk5793, druid:yw448qd5583, druid:ky082np9069, druid:gf424bq2099, druid:rd235fv8920, druid:fz099qm8777, druid:bq782pp3515, druid:jx066kr2212
Error: key not found: "Personal" (11 errors)
Examples: druid:rr760kw8077, druid:sd106bj5619, druid:jc292hs6731, druid:fb460yd3862, druid:jv715vs5567, druid:dr978cj5855, druid:sd744qc5142, druid:mg668kk8822, druid:zj739jx1550, druid:wd641rc1638
Error: key not found: "primary" (3 errors)
Examples: druid:rw667tg8009, druid:md049mq5771, druid:wq793rc8271
Error: key not found: "topic" (1 errors)
Examples: druid:kd333yz5798
Error: key not found: "naf" (1 errors)
Examples: druid:pw260mt1867
Error: key not found: "personal" (1 errors)
Examples: druid:sm835nh3655
Error: key not found: "lcsh" (1 errors)
Examples: druid:xw235wd2174
```

**Comment by justinlittman** [comment]
That's still problematic. I'd expect < 1000, which makes it seem that this is the culprit:
```
Error: #/components/schemas/Standard does not define properties: type (2585 errors)
Examples: druid:zw685dw6356, druid:hv751kj0082, druid:ft545jn7796, druid:pf602td9391, druid:hv022gv3986, druid:wt776rw3893, druid:tt596nz1569, druid:mn953qz8474, druid:mj707sx9198, druid:jr037zj8458
```

**Review by justinlittman — CHANGES_REQUESTED** [review]
Putting a hold on this until the testing issue can be rectified.

**Comment by aaron-collier** [comment]
These appear to be the remaining errors against this branch and not on master:

```
Error: #/components/schemas/DescriptiveBasicValue does not define properties: language (349 errors)
Examples: druid:nh901ng6523, druid:yx344pr0923, druid:xy581jd9710, druid:vv853br8653, druid:tk635kd7152, druid:wb876vv6035, druid:xv866ss4847, druid:xq983jq4604, druid:xt444rj4703, druid:xs653hz4172

37 on master
Error:  isn't one of in #/components/schemas/DescriptiveBasicValue/allOf/2/properties/value (83 errors)
Examples: druid:jp270pd4425, druid:rk160wd1685, druid:tq048zj2285, druid:cg786yh6695, druid:nf221qh5784, druid:dg875gq3366, druid:fd184wn0356, druid:xj910rk4974, druid:zx851yx8537, druid:qj331tp7537



2 on master
Error: #/components/schemas/Source/properties/code does not allow null values (56 errors)
Examples: druid:dv648jc0542, druid:bv130zg8994, druid:dz692fn7184, druid:pc866nj8061, druid:vv180sr0636, druid:sg578sb1573, druid:gd289np1088, druid:pc434cq1315, druid:ms180kr4831, druid:nw905rb5515
```

Working on clearing these up, if not blockers to this PR, I'm happy to put into a follow up or hold on this one for this small set of errors.

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ Metrics/AbcSize:
 # Offense count: 7
 # Configuration parameters: CountComments, CountAsOne.
 Metrics/ClassLength:
-  Max: 118
+  Max: 119
```
Comment (jcoyne) [inline .rubocop_todo.yml:34]: Please wrap the class with an excepetion rather than raising the global threshold.

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: .rubocop_todo.yml (hunk 1 above)
Comment (aaron-collier) [reply .rubocop_todo.yml:34]: Sure, not a fan of that approach myself, but happy to do so if we want to follow that pattern.

Comment (jcoyne) [reply .rubocop_todo.yml]: @aaron-collier The problem with raising the global threshold is that every exception raises the bar, and quality overall devolves.  Another alternative is to refactor the class you're working on so it's not so big.

File: app/services/cocina/from_fedora/descriptive.rb [hunk 2]
```diff
@@ -5,6 +5,30 @@ module FromFedora
...
+      IANA_TERMS = { value: 'IANA media type terms' }.freeze
+      LANGUAGE = { code: 'eng' }.freeze
+      MEDIA_TYPE = 'media type'
+      POINT_COORDS = 'point coordinates'
+      TYPE = 'type'
+
+      # Directional Constants for GEO
+      SOUTH = 'south'
+      WEST = 'west'
+      NORTH = 'north'
```
Comment (jcoyne) [inline app/services/cocina/from_fedora/descriptive.rb:30]: Why are these defined in this class if they're never used by this class?

File: .rubocop_todo.yml (hunk 1 above)
Comment (aaron-collier) [reply .rubocop_todo.yml]: That makes sense, this sort of changed under me, but one line :-), but wrapping is fine here.

File: app/services/cocina/from_fedora/descriptive.rb (hunk 2 above)
Comment (aaron-collier) [reply app/services/cocina/from_fedora/descriptive.rb:30]: I can move them. I was trying to keep the constants together, but that probably doesn't make as much sense as it did to me a couple of weeks ago.

**Review by mjgiarlo — COMMENTED** [review]
Note to reviewers: please squash-merge this when it's ready to go. Thanks!

File: app/services/cocina/from_fedora/descriptive.rb [hunk 3]
```diff
@@ -5,6 +5,10 @@ module FromFedora
     # Creates Cocina Descriptive objects from Fedora objects
     class Descriptive
       DESC_METADATA_NS = Dor::DescMetadataDS::MODS_NS
+      DUBLIN_CORE_NS = 'http://purl.org/dc/elements/1.1/'
+      RDF_NS = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
+      GMD_NS = 'http://www.isotc211.org/2005/gmd'
+      GML_NS = 'http://www.opengis.net/gml/3.2/'
```
Comment (jcoyne) [inline app/services/cocina/from_fedora/descriptive.rb:11]: Can these move into `Geographic` as well?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/services/cocina/to_fedora/descriptive/geographic.rb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        end
+
+        def write
+          return if geo.nil?
+
+          attributes = {}
+          attributes[:displayLabel] = 'geo'
+          xml.extension attributes do
+            xml['rdf'].RDF(format_namespace(geo[:subject].first[:type])) do
+              # REVIST: Need the druid to include rdf:about
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/geographic.rb:32]: Does this need to be done here or ticketed for the future?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/services/cocina/to_fedora/descriptive/geographic.rb (hunk 4 above)
Comment (aaron-collier) [reply app/services/cocina/to_fedora/descriptive/geographic.rb:32]: I think it should be ticketed. I've noted it here, but I think it's outside of the scope of this issue/pr.

File: app/services/cocina/from_fedora/descriptive.rb (hunk 3 above)
Comment (aaron-collier) [reply app/services/cocina/from_fedora/descriptive.rb]: Sure. Done.

File: app/services/cocina/to_fedora/descriptive/geographic.rb (hunk 4 above)
Comment (jcoyne) [reply app/services/cocina/to_fedora/descriptive/geographic.rb:32]: Please make the ticket and then feel free to merge.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**